### PR TITLE
[StorageQuota] Adding StorageQuotaEnforcer and StorageQuotaService

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Complete the implementation for StorageQuotaService

### CODE REVIEW

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaServiceFactory.java
Comment (jsjtzyy) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaServiceFactory.java:46]: I think all of this can be instantiated in `AmbryStorageQuotaService` class. Hence, the constructor of `AmbryStorageQuotaService` can be simplified to:
```
public AmbryStorageQuotaService(StorageQuotaConfig config, StorageQuotaMetrics metrics)
```

File: ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java
Comment (jsjtzyy) [inline ambry-quota/src/main/java/com/github/ambry/quota/HelixStorageUsageRefresher.java:143]: Remove this.

**Review by jsjtzyy — COMMENTED** [review]
Some minor comments. Mostly looks good.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaEnforcer.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaEnforcer.java:60]: yes, it comes from nuage, but it's stored in a different service, called QuotaIn.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java:84]: we will only call it once.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java:95]: we will only call it once.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java:112]: hmm, this would be complicated since storage has to know about the deleted container. I would have to expose another  method in `StorageQuotaService` for that. Will do this later. Thanks for bring this up, will do this later, by adding a TODO .

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaService.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaService.java:40]: it's already used in helixStorageUsageRefresher, passing it to storagequotaservice is to close it when shutting the service down.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaServiceFactory.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaServiceFactory.java:35]: oh, yes, i haven't figured out what metric to add, but just put metric registry here.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java
Comment (ankagrawal) [inline ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java:20]: Its not clear to me whats the difference between `StorageQuotaService` and `StorageQuotaEnforcer`. This interface seems to have same methods as `StorageQuotaEnforcer`.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java
Comment (ankagrawal) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java:88]: It should be this class' responsibility to register listener to `AmbryStorageQuotaService`. Instead of returning `Listener` objects, this class should create the object and `AmbryStorageQuotaService.registerListener(listener)`. This will make things simple, as the only responsibility of `AmbryStorageQuotaService` will be to called listener that are registered by consumers. Also the `AmbryStorageQuotaEnforcer` doesn't need to do anything if it doesn't want to listen.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java
Comment (ankagrawal) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java:99]: javadocs?

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaEnforcer.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaEnforcer.java]: Updated, added a `registerListeners` method in `StorageQuotaEnforcer`.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java
Comment (justinlin-linkedin) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java]: updated.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaEnforcer.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaEnforcer.java:73]: the quota is enforced per container. Different containers would have different quota, thus will be throttled at different time. `QuotaMode` now is a global setting. If we want to stop throttling per container. Then the current interface would have to be changed. I probably would have to revisit the interface when I start implementing for the closed source ambry.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaService.java:20]: StorageQuotaService is more like a wrapper, it's responsibility is to 1) choose different implementation of `StorageQuotaSource` `StorageUsageRefresher` and `StorageQuotaEnforcer` 2) glue them together and make it work.

But mainly the reason why I have `StorageQuotaService` and `StorageQuotaServiceFactory` is to use different implementations of those components. In closed source ambry, I would create a QuotaInStorageQuotaSource, which will get the quota from QuotaIn service, a HelixBasedDataPlanStorageUsageRefresher, which will get the monthly storage usage instead of total usage from helix. And probably a different storage quota enforcer.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaService.java
Comment (jsjtzyy) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaService.java:40]: I understand that. I guess I didn't make it clear previously. The comment is together with the comment in `AmbryStorageQuotaServiceFactory` class where I suggested moving scheduler, quotaSource etc into  `AmbryStorageQuotaService` ctor.  That comment was resolved but I didn't see the change. Did I miss anything?

**Review by jsjtzyy — APPROVED** [review]
LGTM after one minor followup is addressed.

File: ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaEnforcer.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/quota/StorageQuotaEnforcer.java:73]: Nit: maybe name it `setGlobalQuotaMode` to make it explicit.

**Review by ankagrawal — APPROVED** [review]
LGTM. Will wait for test to pass and yingyi's last comment addressed before merging.

**Pull request merged**
